services/download/src/main/java/app/coronawarn/server/services/download/FederationGatewayDownloadService.java

```diff
@@ -0,0 +1,106 @@
+
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
+
+import app.coronawarn.server.common.federation.client.FederationGatewayClient;
+import app.coronawarn.server.common.protocols.external.exposurenotification.DiagnosisKeyBatch;
+import feign.FeignException;
+import java.time.LocalDate;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+
+/**
+ * Responsible for downloading and storing batch information from the federation gateway.
+ */
+@Component
+public class FederationGatewayDownloadService {
+
+  public static final String HEADER_BATCH_TAG = "batchTag";
+  public static final String HEADER_NEXT_BATCH_TAG = "nextBatchTag";
+  public static final String EMPTY_HEADER = "null";
+  private static final Logger logger = LoggerFactory.getLogger(FederationGatewayDownloadService.class);
+  private final FederationGatewayClient federationGatewayClient;
+
+  /**
+   * Constructor.
+   *
+   * @param federationGatewayClient A {@link FederationGatewayClient} for retrieving federation diagnosis key batches.
+   */
+  public FederationGatewayDownloadService(FederationGatewayClient federationGatewayClient) {
+    this.federationGatewayClient = federationGatewayClient;
+  }
+
+  /**
+   * Download the first batch from the EFGS for the given date.
+   *
+   * @param date The date for which the batch should be downloaded.
+   * @return The {@link BatchDownloadResponse} containing the downloaded batch, batchTag and nextBatchTag.
+   */
+  public BatchDownloadResponse downloadBatch(LocalDate date) {
+    try {
+      logger.info("Downloading first batch for date {}", date);
+      ResponseEntity<DiagnosisKeyBatch> response = federationGatewayClient
+          .getDiagnosisKeys(date.format(ISO_LOCAL_DATE));
+      return parseResponseEntity(response);
+    } catch (FeignException e) {
+      throw new FederationGatewayException("Downloading batch for date " + date.format(ISO_LOCAL_DATE) + " failed.", e);
+    }
+  }
+
+  /**
+   * Download the batch from the EFGS with the given batchTag for the given date.
+   *
+   * @param batchTag The batchTag of the batch that should be downloaded.
+   * @param date     The date for which the batch should be downloaded.
+   * @return The {@link BatchDownloadResponse} containing the downloaded batch, batchTag and nextBatchTag.
+   */
+  public BatchDownloadResponse downloadBatch(String batchTag, LocalDate date) {
+    String dateString = date.format(ISO_LOCAL_DATE);
+    try {
+      logger.info("Downloading batch for date {} and batchTag {}.", batchTag, dateString);
+      ResponseEntity<DiagnosisKeyBatch> response = federationGatewayClient
+          .getDiagnosisKeys(batchTag, dateString);
+      return parseResponseEntity(response);
+    } catch (FeignException e) {
+      logger.error("Downloading batch for date {} and batchTag {} failed.", batchTag, dateString);
```

Comment:
Here a failed batch is logged with its batch tag and date. But also where this method is called, the exception handling is also logging the error in a similar way:

**FederationBatchProcessor [138]:**
```java
logger.error("Federation batch processing for date {} and batchTag {} failed. Status set to {}",
          date, batchTag, errorStatus.name(), e);
```

Reply:
The first one is related to the actual download, the second one is for processing it, according to the different layers of FederationGatewayService and FederationBatchProcessor